rdflib/resolver.py

```diff
@@ -0,0 +1,234 @@
+
+This module implements an extensible resolver mechanism that can be used by parsers to
+dereference documents referred to in the process of parsing.
+
+This functionality comes into play with the JSON-LD parser, and is used when resolving
+a document's `@context` document.
+
+A resolver instance can be passed to :meth:`~rdflib.Graph:parse`:
+
+    >>> from rdflib import Graph
+    >>>
+    >>> graph = Graph()
+    >>> graph.parse(
+    ...     data="<ex:s> <ex:p> <ex:o> .",
+    ...     format="application/n-triples",
+    ...     resolver=PermissiveResolver()) # doctest: +ELLIPSIS
+    <Graph identifier=... (<class 'rdflib.graph.Graph'>)>
+
+If no resolver is specified when calling `parse()`, RDFLib falls back to the
+interpreter-wide default resolver. A default resolver instance can be set in code using
+:func:`set_default_resolver`, or via the `RDFLIB_DEFAULT_RESOLVER_CLASS` environment
+variable using a colon-delimited style (i.e. `"path.to.module:ClassName"`). The current
+default resolver instance can be retrieved using :func:`get_default_resolver`.
+
+If the default resolver is not explicitly set, the default resolver will be an instance
+of :class:`DefaultResolver`, which is implements a default-deny policy. This policy can
+be altered using the `RDFLIB_RESOLVABLE_URL_SCHEMES` and
+`RDFLIB_RESOLVABLE_URL_ALLOWLIST` environment variables. The first should be a
+whitespace-delimited list of URL schemes (e.g. `"file https"`) and the second should be a
+whitespace-delimited list of allowed URLs.
+
+Custom resolvers should subclass :class:`Resolver`, e.g.:
+
+    >>> class LocalResolver(Resolver):
+    ...     documents = {}
+    ...
+    ...     def is_resolution_allowed(self, scheme: str, url: str) -> bool:
+    ...         return url in documents or super().is_resolution_allowed(scheme, url)
+    ...
+    ...     @url_resolver(schemes={"http", "https"})
+    ...     def resolve_http(self, url: str, format: str, scheme: str) -> InputSource:
+    ...         return self.documents[url]
+
+To replicate the permissive behaviour of older releases, you can set
+:class:`PermissiveResolver` as your default resolver.
+"""
+
+import functools
+import importlib
+import os
+import pathlib
+import threading
+from urllib.parse import urlparse
+from urllib.request import url2pathname
+
+from .exceptions import ResolutionError, ResolutionForbiddenError
+from .parser import FileInputSource, InputSource, URLInputSource
+from .term import URIRef
+
+
+def url_resolver(schemes):
+    """Decorator for Resolver methods that resolve URLs into InputSources.
+
+    :param schemes: A collection of URI schemes as str objects that this resolver
+        provides.
+    """
+
+    def wrapper(func):
+        func._url_resolver_schemes = schemes
+        return func
+
+    return wrapper
+
+
+DEFAULT_RESOLVABLE_URL_ALLOWLIST = {
+    "https://w3c.github.io/json-ld-rc/context.jsonld",
+}
+
+_unspecified = object()
+
+
+class Resolver:
```

Comment:
There should be an abstract base class for this so someone can implement it without all the baggage from the default implementation.